zerver/lib/upload.py

```diff
@@ -322,33 +311,51 @@ def get_file_info(request: HttpRequest, user_file: File) -> Tuple[str, int, Opti
 
 
 def get_signed_upload_url(path: str) -> str:
-    conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
-    return conn.generate_url(15, 'GET', bucket=settings.S3_AUTH_UPLOADS_BUCKET, key=path)
+    client = boto3.client(
+        's3',
+        aws_access_key_id=settings.S3_KEY,
+        aws_secret_access_key=settings.S3_SECRET_KEY)
+    params = {
+        'Bucket': settings.S3_AUTH_UPLOADS_BUCKET,
+        'Key': path}
+    return client.generate_presigned_url(
+        ClientMethod='get_object',
+        Params=params,
+        ExpiresIn=15,
+        HttpMethod='GET')
 
 def get_realm_for_filename(path: str) -> Optional[int]:
-    conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
-    key = get_bucket(conn, settings.S3_AUTH_UPLOADS_BUCKET).get_key(path)  # type: Optional[Key]
-    if key is None:
-        # This happens if the key does not exist.
+    try:
+        s3_object = get_s3_object(settings.S3_AUTH_UPLOADS_BUCKET, path)
+        user_profile_id = s3_object.metadata['user_profile_id']
+    except Exception:
+        # Boto3 has weird exception handling that doesn't
+        # work that reliably
+        # https://github.com/boto/boto3/issues/1195
         return None
-    return get_user_profile_by_id(key.metadata["user_profile_id"]).realm_id
+    return get_user_profile_by_id(user_profile_id).realm_id
 
 class S3UploadBackend(ZulipUploadBackend):
     def __init__(self) -> None:
-        self.connection = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
+        self.session = boto3.Session(settings.S3_KEY, settings.S3_SECRET_KEY,
+                                     region_name='s3.amazonaws.com')
```

Comment:
I'm not sure if setting `region_name` is the right call here.  But it appears boto3 doesn't have a `DefaultHost` equivalent in `Session`.  As it seems it now requires you to specify the default.  Which we can do here, and then an `aws_config` could override that if needed I'm guessing? Not sure.